README.md
# react-leaflet-google-v2 [![npm version](https://img.shields.io/npm/v/react-leaflet-google-v2.svg)](https://www.npmjs.com/package/react-leaflet-google-v2)![npm downloads](https://img.shields.io/npm/dw/react-leaflet-google-v2)[![Node.js CI](https://github.com/TA-Geoforce/react-leaflet-google-v2/actions/workflows/node.js.yml/badge.svg)](https://github.com/TA-Geoforce/react-leaflet-google-v2/actions/workflows/node.js.yml)[![Release](https://github.com/TA-Geoforce/react-leaflet-google-v2/actions/workflows/npm-publish.yml/badge.svg)](https://github.com/TA-Geoforce/react-leaflet-google-v2/actions/workflows/npm-publish.yml)

Most recently tested with Leaflet React-Leaflet 3.0.5.

GoogleMaps layer as React component for Leaflet build on top of [React-Leaflet](https://github.com/PaulLeCam/react-leaflet). 

The google maps layer is using the plugin from [Leaflet.GridLayer.GoogleMutant](https://gitlab.com/IvanSanchez/Leaflet.GridLayer.GoogleMutant) 

Also it uses [google-maps](https://www.npmjs.com/package/google-maps), a wrapper for asynchronously download Google Maps API in the browser.

![Example](images/example.gif)

## Complete example with react-leaflet

To get started, to be able to run the example you have execute:
```bash
npm install
```

```jsx
import React from 'react';
import { MapContainer, TileLayer, LayersControl } from 'react-leaflet'
import { GoogleLayer } from '../src'
const { BaseLayer } = LayersControl;
const key = 'your key goes here';
const terrain = 'TERRAIN';
const road = 'ROADMAP';
const satellite = 'SATELLITE';
const hydrid = 'HYBRID';
//// Google's map type. Valid values are 'roadmap', 'satellite' or 'terrain'. 'hybrid' is not really supported.

export default class GoogleExample extends React.Component {

  constructor() {
    super();
  }

  render() {
    return (
      <MapContainer center={[42.09618442380296, -71.5045166015625]} zoom={2} zoomControl={true}>
        <LayersControl position='topright'>
          <BaseLayer  name='OpenStreetMap.Mapnik'>
            <TileLayer  url="http://{s}.tile.osm.org/{z}/{x}/{y}.png"/>
          </BaseLayer>
         <BaseLayer checked name='Google Maps Roads'>
            <GoogleLayer googlekey={key}  maptype={road} />
          </BaseLayer>
         <BaseLayer  name='Google Maps Terrain'>
            <GoogleLayer googlekey={key}  maptype={terrain} />
          </BaseLayer>
           <BaseLayer  name='Google Maps Satellite'>
            <GoogleLayer googlekey={key}  maptype={satellite} />
          </BaseLayer>
            <BaseLayer  name='Google Maps Hydrid'>
            <GoogleLayer googlekey={key}  maptype={hydrid}  libraries={['geometry', 'places']} />
          </BaseLayer>  
          <BaseLayer  name='Google Maps with Libraries'>
            <GoogleLayer googlekey={key}  maptype={hydrid}  libraries={['geometry', 'places']} />
          </BaseLayer>        
        </LayersControl>
      </MapContainer>
    )
  }
}


```

You can find the following example in the folder ```example```. Run the above code by executing the following scripts in package.json, with the order they are stated underneath:

1) ```build```

1) ```example```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

### [v6.0.0](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.2.0...v6.0.0)

> 17 February 2022

- [skip-ci] [`e7bfb1d`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/e7bfb1d187710122a3ac866dc913706a7857b4b0)
- fix: fix map not loading [`98c2c9b`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/98c2c9bebb2eb0709b3ebed21a122612873ce699)

#### [v5.2.0](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.11...v5.2.0)

> 7 July 2021

- feat: fix paths [`#25`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/25)
- [skip-ci] [`d03eb30`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/d03eb309a0d64c6fb47b0476a3d041aef2864263)

#### [v5.1.11](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.10...v5.1.11)

> 7 July 2021

- fix: fixed semantic release scipts for prepublish [`#24`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/24)
- [skip-ci] [`ed7a3e5`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/ed7a3e56a122910a4423eac362bc1299bd40ae54)

#### [v5.1.10](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.9...v5.1.10)

> 7 July 2021

- fix: fix dependecies [`#23`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/23)
- fix: fix package [`9b773c6`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/9b773c6b48577a0110651809b0be0eb622ff5817)
- [skip-ci] [`8780691`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/8780691dd928500cead223bf8c618458588e05f2)

#### [v5.1.9](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.8...v5.1.9)

> 7 July 2021

- fix: deleted webpack generated files [`#21`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/21)
- [skip-ci] [`78313ce`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/78313ce61d22d1e7604926681ec844111dd2fc0f)

#### [v5.1.8](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.7...v5.1.8)

> 7 July 2021

- fix: npmignore fix [`#20`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/20)
- [skip-ci] [`dd8423a`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/dd8423ac09635d1ebb557262b857a9526a333190)

#### [v5.1.7](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.6...v5.1.7)

> 7 July 2021

- fix: fix-minified-error [`#19`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/19)
- [skip-ci] [`aeba20d`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/aeba20dfe16a2fd181c5b4dfd9c73d1c38d7b945)

#### [v5.1.6](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.5...v5.1.6)

> 7 July 2021

- fix: added missing library in package [`#18`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/18)
- [skip-ci] [`99d0609`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/99d0609a5940f997fcd0cf399521d1a8be62bf11)

#### [v5.1.5](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.4...v5.1.5)

> 14 June 2021

- fix: package lock [`#15`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/15)
- fix: package [`#14`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/14)

#### [v5.1.4](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.3...v5.1.4)

> 14 June 2021

- fix: publish [`#13`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/13)
- Update README.md [`#12`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/12)
- Update package.json [`#11`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/11)
- added semantic release code and created PAT [`#10`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/10)
- Create npm-publish.yml [`#9`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/9)
- Feature/upgrade to react leaflet v3 [`#7`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/7)
- Create node.js.yml [`#8`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/8)
- changes for react-leaflet v3 [`#5`](https://github.com/TA-Geoforce/react-leaflet-google-v2/pull/5)
- Delete .github/workflows directory [`5f5597b`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/5f5597b03c50f26c8cf27bb280440a1c2de894bf)
- Create node_publish.js.yml [`92677b1`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/92677b155b01f362f90e7119a64af010f5a07c56)
- Delete node_publish.js.yml [`12ae76e`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/12ae76efbd4217f4e825390ee96069781b50f7d6)

#### [v5.1.3](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.2...v5.1.3)

> 13 November 2019

- change main npm [`4577482`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/457748204be1836d542940c8dd33da46701fc075)

#### [v5.1.2](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.1...v5.1.2)

> 13 November 2019

- fix webpack issue [`ca75169`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/ca751696921e6f822b9ea8ecbac460ce7f18f120)

#### [v5.1.1](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.1.0...v5.1.1)

> 13 November 2019

- fix dev dependencies [`30c253f`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/30c253f5024e9d700c6ab773221ab24b8bbf19bf)

#### [v5.1.0](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.0.1...v5.1.0)

> 13 November 2019

- fix dist version [`e6f8bd6`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/e6f8bd6b709b0d2991f129792b0406cfa1c99f25)
- clean package [`6aee721`](https://github.com/TA-Geoforce/react-leaflet-google-v2/commit/6aee721e4ba8b07830458bfdc2617662059438b6)

#### [v5.0.1](https://github.com/TA-Geoforce/react-leaflet-google-v2/compare/v5.0.0...v5.0.1)

> 6 November 2019


__tests__/GoogleLayer.test.js
import React from 'react';
import { MapContainer, LayersControl } from 'react-leaflet'
import { GoogleLayer } from '../src'
import { mount } from "enzyme"
import Enzyme from "enzyme"
import Adapter from "@wojtekmaj/enzyme-adapter-react-17"

Enzyme.configure({ adapter: new Adapter() })

const { BaseLayer } = LayersControl;

const props = {
    googlekey: "test-key",
    maptype: "ROADMAP",
}
describe("Check if Google Layer renders", () => {
    it("Simple render of component", () => {
        const wrapper = mount(
            <MapContainer center={[42.09618442380296, -71.5045166015625]} zoom={2} zoomControl={true}>
                <LayersControl position='topright'>
                    <BaseLayer  name='Google Maps with Libraries'>
                        <GoogleLayer {...props} />
                    </BaseLayer>
                </LayersControl>
            </MapContainer>
        )
        expect((wrapper).find(GoogleLayer).prop('googlekey')).toEqual("test-key");
        expect((wrapper).find(GoogleLayer).prop('maptype')).toEqual("ROADMAP");
    })
})

webpack.common.js
/* eslint-disable */
module.exports = {
    entry: './dist/react-leaflet-google-v2.min.js',
    output: {
        library: {
            root: 'ReactLeafletGoogle',
            amd: 'react-leaflet-google-v2',
            commonjs: 'react-leaflet-google-v2'
        },
        libraryExport: 'default',
        libraryTarget: 'umd'
    },
    externals: {
        debug: 'debug',
        leaflet: {
            commonjs: 'leaflet',
            commonjs2: 'leaflet',
            root: 'L'
        },
        'react-leaflet': {
            commonjs: 'react-leaflet',
            commonjs2: 'react-leaflet',
            root: 'ReactLeaflet'
        },
        react: {
            commonjs: 'react',
            commonjs2: 'react',
            root: 'React'
        }
    },
    mode: 'production',
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                exclude: /(node_modules|bower_components)/,
                loader: 'babel-loader',
                options: {
                    presets: ['env', 'react'],
                    plugins: ['transform-class-properties']
                }
            },
            {
                test: /\.css$/,
                exclude: /node_modules/,
                use: [
                    { loader: 'style-loader' },
                    { loader: 'css-loader' }
                ]
            }
        ]
    }
};

webpack.config.js
/* eslint-disable */
module.exports = {
  output: {
    library: 'ReactLeaflet',
    libraryTarget: 'umd'
  },
  externals: [
    {
      leaflet: {
        amd: 'leaflet',
        commonjs: 'leaflet',
        commonjs2: 'leaflet',
        root: 'L'
      },
    },
    {
      'react-leaflet': {
        amd: 'react-leaflet',
        commonjs: 'react-leaflet',
        commonjs2: 'react-leaflet'
      }
    },
    {
      react: {
        root: 'React',
        commonjs2: 'react',
        commonjs: 'react',
        amd: 'react'
      }
    },
    {
      'google-maps': {
        amd: 'google-maps',
        commonjs: 'google-maps',
        commonjs2: 'google-maps',
        root: 'GoogleMapsLoader'
      }
    }
  ],
  mode: "production",
  module: {
		rules: [
            {
                test: /\.(js|jsx)$/,
				use: {
					loader: "babel-loader",
					options: {
					  presets: ["@babel/preset-env"],
					}
				  },
            },
            {
                test: /\.css$/,
                use: ["style-loader", "css-loader"]
            }
        ]
	}
};



src/leaflet.google.js
// Based on https://github.com/shramov/leaflet-plugins
// GridLayer like https://avinmathew.com/leaflet-and-google-maps/ , but using MutationObserver instead of jQuery
import GoogleMapsLoader from "google-maps";
import * as L from "leaflet";



let google;

// class GridLayer.GoogleMutant
// extends GridLayer
L.GridLayer.GoogleMutant = L.GridLayer.extend({
  includes: L.Evented,
  
  options: {
    minZoom: 0,
    maxZoom: 21,
    tileSize: 256,
    subdomains: "abc",
    errorTileUrl: "",
    attribution: "", // The mutant container will add its own attribution anyways.
    opacity: 1,
    continuousWorld: false,
    noWrap: false,
    // 🍂option type: String = 'roadmap'
    // Google's map type. Valid values are 'roadmap', 'satellite' or 'terrain'. 'hybrid' is not really supported.
    type: "HYBRID",
    maxNativeZoom: 21,
  },
  
  initialize(options) {
    L.GridLayer.prototype.initialize.call(this, options);
    let self = this;
    GoogleMapsLoader.VERSION = options.version || "3.43";
    if ( options.asclientid ) {
      GoogleMapsLoader.CLIENT = options.googlekey;
    } else {
      GoogleMapsLoader.KEY = options.googlekey;
    }
    
    GoogleMapsLoader.LIBRARIES = options.libraries || [];
    
    if ( options.channel ) {
      GoogleMapsLoader.CHANNEL = options.channel;
    }
    
    self._type = options.maptype || "SATELLITE";
    
    GoogleMapsLoader.load(function(_google) {
      google = _google;
      self._ready = true;
      // self._initMapObject();
      // self.
      self._initMutant();
      self._update();
      
      if ( options.onAfterLoad ) {
        options.onAfterLoad(google);
      }
      
      // this._ready = google.maps.Map !== undefined;
      // if (!this._ready) L.Google.asyncWait.push(this);
    });
    
    // Couple data structures indexed by tile key
    this._tileCallbacks = {}; // Callbacks for promises for tiles that are expected
    this._freshTiles = {}; // Tiles from the mutant which haven't been requested yet
    
    this._imagesPerTile = self._type === "HYBRID" ? 2 : 1;
    this.createTile =
      self._type === "HYBRID" ? this._createMultiTile : this._createSingleTile;
  },
  
  onAdd(map) {
    L.GridLayer.prototype.onAdd.call(this, map);
    this._initMutantContainer();
    
    // Will be called after callback passed in initialize function
    GoogleMapsLoader.load(
      function(google) {
        this._map = map;
        
        this._initMutant();
        
        map.on("viewreset", this._reset, this);
        map.on("move", this._update, this);
        map.on("zoomend", this._handleZoomAnim, this);
        map.on("resize", this._resize, this);
        
        // 20px instead of 1em to avoid a slight overlap with google's attribution
        if ( map._controlCorners )
          map._controlCorners.bottomright.style.marginBottom = "20px";
        
        this._reset();
        this._update();
      }.bind(this)
    );
  },
  
  onRemove(map) {
    L.GridLayer.prototype.onRemove.call(this, map);
    map._container.removeChild(this._mutantContainer);
    this._mutantContainer = undefined;
    
    map.off("viewreset", this._reset, this);
    map.off("move", this._update, this);
    map.off("zoomend", this._handleZoomAnim, this);
    map.off("resize", this._resize, this);
    
    if ( map._controlCorners )
      map._controlCorners.bottomright.style.marginBottom = "0em";
  },
  
  getAttribution() {
    return this.options.attribution;
  },
  
  setOpacity(opacity) {
    this.options.opacity = opacity;
    if ( opacity < 1 ) {
      L.DomUtil.setOpacity(this._mutantContainer, opacity);
    }
  },
  
  setElementSize(e, size) {
    e.style.width = size.x + "px";
    e.style.height = size.y + "px";
  },
  
  _initMutantContainer() {
    if ( !this._mutantContainer ) {
      this._mutantContainer = L.DomUtil.create(
        "div",
        "leaflet-google-mutant leaflet-top leaflet-left"
      );
      this._mutantContainer.id =
        "_MutantContainer_" + L.Util.stamp(this._mutantContainer);
      // 			this._mutantContainer.style.zIndex = 'auto';
      this._mutantContainer.style.pointerEvents = "none";
      
      this._map.getContainer().appendChild(this._mutantContainer);
    }
    
    this.setOpacity(this.options.opacity);
    this.setElementSize(this._mutantContainer, this._map.getSize());
    
    this._attachObserver(this._mutantContainer);
  },
  
  _initMutant() {
    if ( !this._ready || !this._mutantContainer ) return;
    this._mutantCenter = new google.maps.LatLng(0, 0);
    
    let map = new google.maps.Map(this._mutantContainer, {
      center: this._mutantCenter,
      zoom: 0,
      tilt: 0,
      mapTypeId: google.maps.MapTypeId[ this._type ],
      disableDefaultUI: true,
      keyboardShortcuts: false,
      draggable: false,
      disableDoubleClickZoom: true,
      scrollwheel: false,
      streetViewControl: false,
      styles: this.options.styles || [],
      backgroundColor: "transparent",
    });
    
    this._mutant = map;
    
    // 🍂event spawned
    // Fired when the mutant has been created.
    this.fire("spawned", {
      mapObject: map,
    });
  },
  
  _attachObserver: function _attachObserver(node) {
    // 		console.log('Gonna observe', node);
    
    let observer = new MutationObserver(this._onMutations.bind(this));
    
    // pass in the target node, as well as the observer options
    observer.observe(node, {
      childList: true,
      subtree: true,
    });
  },
  
  _onMutations: function _onMutations(mutations) {
    for ( let i = 0; i < mutations.length; ++i ) {
      let mutation = mutations[ i ];
      for ( let j = 0; j < mutation.addedNodes.length; ++j ) {
        let node = mutation.addedNodes[ j ];
        
        if ( node instanceof HTMLImageElement ) {
          this._onMutatedImage(node);
        } else if ( node instanceof HTMLElement ) {
          Array.prototype.forEach.call(
            node.querySelectorAll("img"),
            this._onMutatedImage.bind(this)
          );
        }
      }
    }
  },
  
  // Only images which 'src' attrib match this will be considered for moving around.
  // Looks like some kind of string-based protobuf, maybe??
  // Only the roads (and terrain, and vector-based stuff) match this pattern
  _roadRegexp: /!1i(\d+)!2i(\d+)!3i(\d+)!/,
  
  // On the other hand, raster imagery matches this other pattern
  _satRegexp: /x=(\d+)&y=(\d+)&z=(\d+)/,
  
  // On small viewports, when zooming in/out, a static image is requested
  // This will not be moved around, just removed from the DOM.
  _staticRegExp: /StaticMapService\.GetMapImage/,
  
  _onMutatedImage: function _onMutatedImage(imgNode) {
    // 		if (imgNode.src) {
    // 			console.log('caught mutated image: ', imgNode.src);
    // 		}
    
    let coords;
    let match = imgNode.src.match(this._roadRegexp);
    let sublayer;
    let parent;
    
    if ( match ) {
      coords = {
        z: match[ 1 ],
        x: match[ 2 ],
        y: match[ 3 ],
      };
      if ( this._imagesPerTile > 1 ) {
        imgNode.style.zIndex = 1;
      }
      sublayer = 1;
    } else {
      match = imgNode.src.match(this._satRegexp);
      if ( match ) {
        coords = {
          x: match[ 1 ],
          y: match[ 2 ],
          z: match[ 3 ],
        };
      }
      // 			imgNode.style.zIndex = 0;
      sublayer = 0;
    }
    
    if ( coords ) {
      let key = this._tileCoordsToKey(coords);
      imgNode.style.position = "absolute";
      if ( this._imagesPerTile > 1 ) {
        key += "/" + sublayer;
      }
      if ( key in this._tileCallbacks && this._tileCallbacks[ key ] ) {
        // console.log('Fullfilling callback ', key);
        this._tileCallbacks[ key ].pop()(imgNode);
        if ( !this._tileCallbacks[ key ].length ) {
          delete this._tileCallbacks[ key ];
        }
      } else {
        // console.log('Caching for later', key);
        parent = imgNode.parentNode;
        if ( parent ) {
          parent.removeChild(imgNode);
          parent.removeChild = L.Util.falseFn;
          // 					imgNode.parentNode.replaceChild(L.DomUtil.create('img'), imgNode);
        }
        if ( key in this._freshTiles ) {
          this._freshTiles[ key ].push(imgNode);
        } else {
          this._freshTiles[ key ] = [ imgNode ];
        }
      }
    } else if ( imgNode.src.match(this._staticRegExp) ) {
      parent = imgNode.parentNode;
      if ( parent ) {
        // Remove the image, but don't store it anywhere.
        // Image needs to be replaced instead of removed, as the container
        // seems to be reused.
        imgNode.parentNode.replaceChild(L.DomUtil.create("img"), imgNode);
      }
    }
  },
  
  // This will be used as this.createTile for 'roadmap', 'sat', 'terrain'
  _createSingleTile: function createTile(coords, done) {
    let key = this._tileCoordsToKey(coords);
    // console.log('Need:', key);
    
    if ( key in this._freshTiles ) {
      let tile = this._freshTiles[ key ].pop();
      if ( !this._freshTiles[ key ].length ) {
        delete this._freshTiles[ key ];
      }
      L.Util.requestAnimFrame(done);
      // 			console.log('Got ', key, ' from _freshTiles');
      return tile;
    }
    let tileContainer = L.DomUtil.create("div");
    this._tileCallbacks[ key ] = this._tileCallbacks[ key ] || [];
    this._tileCallbacks[ key ].push(
      (function(c /* , k*/) {
        return function(imgNode) {
          let parent = imgNode.parentNode;
          if ( parent ) {
            parent.removeChild(imgNode);
            parent.removeChild = L.Util.falseFn;
            // 						imgNode.parentNode.replaceChild(L.DomUtil.create('img'), imgNode);
          }
          c.appendChild(imgNode);
          done();
          // 					console.log('Sent ', k, ' to _tileCallbacks');
        };
      })(tileContainer /* , key*/)
    );
    
    return tileContainer;
  },
  
  // This will be used as this.createTile for 'hybrid'
  _createMultiTile: function createTile(coords, done) {
    let key = this._tileCoordsToKey(coords);
    
    let tileContainer = L.DomUtil.create("div");
    tileContainer.dataset.pending = this._imagesPerTile;
    
    for ( let i = 0; i < this._imagesPerTile; i++ ) {
      let key2 = key + "/" + i;
      if ( key2 in this._freshTiles ) {
        tileContainer.appendChild(this._freshTiles[ key2 ].pop());
        if ( !this._freshTiles[ key2 ].length ) {
          delete this._freshTiles[ key2 ];
        }
        tileContainer.dataset.pending--;
        // 				console.log('Got ', key2, ' from _freshTiles');
      } else {
        this._tileCallbacks[ key2 ] = this._tileCallbacks[ key2 ] || [];
        this._tileCallbacks[ key2 ].push(
          (function(c /* , k2*/) {
            return function(imgNode) {
              let parent = imgNode.parentNode;
              if ( parent ) {
                parent.removeChild(imgNode);
                parent.removeChild = L.Util.falseFn;
                // 							imgNode.parentNode.replaceChild(L.DomUtil.create('img'), imgNode);
              }
              c.appendChild(imgNode);
              c.dataset.pending--;
              if ( !parseInt(c.dataset.pending) ) {
                done();
              }
              // 						console.log('Sent ', k2, ' to _tileCallbacks, still ', c.dataset.pending, ' images to go');
            };
          })(tileContainer /* , key2*/)
        );
      }
    }
    
    if ( !parseInt(tileContainer.dataset.pending) ) {
      L.Util.requestAnimFrame(done);
    }
    return tileContainer;
  },
  
  _checkZoomLevels() {
    // setting the zoom level on the Google map may result in a different zoom level than the one requested
    // (it won't go beyond the level for which they have data).
    // verify and make sure the zoom levels on both Leaflet and Google maps are consistent
    if (
      this._map.getZoom() !== undefined &&
      this._mutant.getZoom() !== this._map.getZoom()
    ) {
      // zoom levels are out of sync. Set the leaflet zoom level to match the google one
      this._map.setZoom(this._mutant.getZoom());
    }
  },
  
  _reset() {
    this._initContainer();
  },
  
  _update() {
    L.GridLayer.prototype._update.call(this);
    if ( !this._mutant ) return;
    
    let center = this._map.getCenter();
    let _center = new google.maps.LatLng(center.lat, center.lng);
    
    this._mutant.setCenter(_center);
    let zoom = this._map.getZoom();
    if ( zoom !== undefined ) {
      this._mutant.setZoom(Math.round(this._map.getZoom()));
    }
  },
  
  _resize() {
    let size = this._map.getSize();
    if (
      this._mutantContainer.style.width === size.x &&
      this._mutantContainer.style.height === size.y
    )
      return;
    this.setElementSize(this._mutantContainer, size);
    if ( !this._mutant ) return;
    google.maps.event.trigger(this._mutant, "resize");
  },
  
  _handleZoomAnim() {
    let center = this._map.getCenter();
    let _center = new google.maps.LatLng(center.lat, center.lng);
    
    this._mutant.setCenter(_center);
    this._mutant.setZoom(Math.round(this._map.getZoom()));
  },
  
  // Agressively prune _freshtiles when a tile with the same key is removed,
  // this prevents a problem where Leaflet keeps a loaded tile longer than
  // GMaps, so that GMaps makes two requests but Leaflet only consumes one,
  // polluting _freshTiles with stale data.
  _removeTile(key) {
    if ( this._imagesPerTile > 1 ) {
      for ( let i = 0; i < this._imagesPerTile; i++ ) {
        let key2 = key + "/" + i;
        if ( key2 in this._freshTiles ) {
          delete this._freshTiles[ key2 ];
        }
        // 				console.log('Pruned spurious hybrid _freshTiles');
      }
    } else if ( key in this._freshTiles ) {
      delete this._freshTiles[ key ];
      // 				console.log('Pruned spurious _freshTiles', key);
    }
    
    return L.GridLayer.prototype._removeTile.call(this, key);
  },
});

// 🍂factory gridLayer.googleMutant(options)
// Returns a new `GridLayer.GoogleMutant` given its options
L.gridLayer.googleMutant = function(options) {
  return new L.GridLayer.GoogleMutant(options);
};

example/google.js
import React from 'react';
import { MapContainer, TileLayer, LayersControl } from 'react-leaflet'
import { GoogleLayer } from '../src/index.js'
const { BaseLayer } = LayersControl;
const key = 'AIzaSyDEG4lyorD61vnJoAHG0FkQERZ-McElZyg';
const terrain = 'TERRAIN';
const road = 'ROADMAP';
const satellite = 'SATELLITE';
const hydrid = 'HYBRID';

export default class GoogleExample extends React.Component {

  constructor() {
    super();
  }

  render() {
    return (
      <MapContainer center={[42.09618442380296, -71.5045166015625]} zoom={2} zoomControl={true}>
        <LayersControl position='topright'>
          <BaseLayer  name='OpenStreetMap.Mapnik'>
            <TileLayer  url="http://{s}.tile.osm.org/{z}/{x}/{y}.png"/>
          </BaseLayer>
         <BaseLayer checked name='Google Maps Roads'>
            <GoogleLayer googlekey={key}  maptype={road} />
          </BaseLayer>
         <BaseLayer  name='Google Maps Terrain'>
            <GoogleLayer googlekey={key}  maptype={terrain} />
          </BaseLayer>
           <BaseLayer  name='Google Maps Satellite'>
            <GoogleLayer googlekey={key}  maptype={satellite} />
          </BaseLayer>
            <BaseLayer  name='Google Maps Hydrid'>
            <GoogleLayer googlekey={key}  maptype={hydrid}  libraries={['geometry', 'places']} />
          </BaseLayer>  
          <BaseLayer  name='Google Maps with Libraries'>
            <GoogleLayer googlekey={key}  maptype={hydrid}  libraries={['geometry', 'places']} />
          </BaseLayer>        
        </LayersControl>
      </MapContainer>
    )
  }
}